arangod/Aql/OutputAqlItemRow.h

```diff
@@ -329,17 +359,55 @@ class OutputAqlItemRow {
     TRI_ASSERT(_block != nullptr);
     return *_block;
   }
+  template <class ItemRowType>
+  inline void doCopyRow(ItemRowType const& sourceRow, bool ignoreMissing);
+
+  template <class ItemRowType>
+  inline void memorizeRow(ItemRowType const& sourceRow);
+
+  template <class ItemRowType>
+  inline bool testIfWeMostClone(ItemRowType const& sourceRow) const;
 
-  inline void doCopyRow(InputAqlItemRow const& sourceRow, bool ignoreMissing);
+  template <class ItemRowType>
+  inline void adjustShadowRowDepth(ItemRowType const& sourceRow);
 };
 
+template <>
+inline void OutputAqlItemRow::memorizeRow<InputAqlItemRow>(InputAqlItemRow const& sourceRow) {
+  _lastSourceRow = sourceRow;
+  _lastBaseIndex = _baseIndex;
+}
+
+template <>
+inline void OutputAqlItemRow::memorizeRow<ShadowAqlItemRow>(ShadowAqlItemRow const& sourceRow) {
+}
+
+template <>
+inline bool OutputAqlItemRow::testIfWeMostClone<InputAqlItemRow>(InputAqlItemRow const& sourceRow) const {
+  return _baseIndex == 0 || _lastSourceRow != sourceRow;
+}
 
-void OutputAqlItemRow::doCopyRow(InputAqlItemRow const& sourceRow, bool ignoreMissing) {
+template <>
+inline bool OutputAqlItemRow::testIfWeMostClone<ShadowAqlItemRow>(ShadowAqlItemRow const& sourceRow) const {
+  return true;
+}
+
+template <>
+inline void OutputAqlItemRow::adjustShadowRowDepth<InputAqlItemRow>(InputAqlItemRow const& sourceRow) {
+}
+
+template <>
+inline void OutputAqlItemRow::adjustShadowRowDepth<ShadowAqlItemRow>(ShadowAqlItemRow const& sourceRow) {
+  block().setShadowRowDepth(_baseIndex, sourceRow.getShadowDepthValue());
+}
+
+template <class ItemRowType>
+inline void OutputAqlItemRow::doCopyRow(ItemRowType const& sourceRow, bool ignoreMissing) {
   // Note that _lastSourceRow is invalid right after construction. However, when
   // _baseIndex > 0, then we must have seen one row already.
   TRI_ASSERT(!_doNotCopyInputRow);
-  TRI_ASSERT(_baseIndex == 0 || _lastSourceRow.isInitialized());
-  bool mustClone = _baseIndex == 0 || _lastSourceRow != sourceRow;
+  bool mustClone = testIfWeMostClone(sourceRow);
+  // TRI_ASSERT(_baseIndex == 0 || _lastSourceRow.isInitialized());
```

Comment:
Is this assertion no longer valid? It's not obvious to me why not, but if it is not, remove the line. Otherwise, the comment. ;)

Reply:
Not valid anymore.
As the last row can be shadow, => not initialized even if baseIndex != 0